### PR TITLE
Add helper objects for decoding UEFI structures around FMP capsules.

### PR DESCRIPTION
**Pull request by corthon**

### CODE REVIEW

**Comment by corthon** [comment]
Do I need to update a readme somewhere?

**Comment by corthon** [comment]
Ooof. That's a lotta flake failures. I'll look into them.

File: edk2toollib/uefi/fmp_auth_header.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    _StructAuthInfoSize = struct.calcsize(_StructAuthInfoFormat)
+
+    _WIN_CERT_REVISION = 0x0200
+    _WIN_CERT_TYPE_EFI_GUID = 0x0EF1
+    _EFI_CERT_TYPE_PKCS7_GUID = uuid.UUID('4aafd29d-68df-49ee-8aa9-347d375665a7')
+
+    def __init__(self):
+        self._Valid = False
+        self.MonotonicCount = 0
+        self.dwLength = self._StructAuthInfoSize
```
Comment (spbrogan) [inline edk2toollib/uefi/fmp_auth_header.py]: this data is actually a wincert which we have defined
https://github.com/tianocore/edk2-pytool-library/blob/master/edk2toollib/uefi/wincert.py#L157

File: edk2toollib/uefi/fmp_auth_header.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    _StructFormat = '<QIHH16s'
+    _StructSize = struct.calcsize(_StructFormat)
+
+    _MonotonicCountFormat = '<Q'
+    _MonotonicCountSize = struct.calcsize(_MonotonicCountFormat)
+
+    _StructAuthInfoFormat = '<IHH16s'
+    _StructAuthInfoSize = struct.calcsize(_StructAuthInfoFormat)
+
+    _WIN_CERT_REVISION = 0x0200
```
Comment (spbrogan) [inline edk2toollib/uefi/fmp_auth_header.py]: defined in https://github.com/tianocore/edk2-pytool-library/blob/master/edk2toollib/uefi/wincert.py#L163

File: edk2toollib/uefi/edk2/fmp_payload_header.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    def Encode(self):
+        FmpPayloadHeader = struct.pack(
+            self._StructFormat,
+            self.Signature,
+            self.HeaderSize,
+            self.FwVersion,
+            self.LowestSupportedVersion
+        )
+        self._Valid = True
```
[the remark this answers is not in the document]
Comment (corthon) [reply edk2toollib/uefi/edk2/fmp_payload_header.py]: Not my code, I'll revise in the next "functional" PR. This was mostly get get parity.

File: edk2toollib/uefi/wincert.py [hunk 4]
```diff
@@ -81,77 +141,181 @@ def Write(self, fs):
...
-        print("  Hdr_wRevision:        0x%X" % self.Hdr_wRevision)
-        print("  Hdr_wCertificateType: 0x%X" % self.Hdr_wCertificateType)
-        print("  CertType:             %s" % str(self.CertType))
-        print("  CertData:             ")
-        cdl = self.CertData.tolist()
-        PrintByteList(cdl)
+        self.DumpInfo()
+
+    def DumpInfo(self):
+        print('EFI_FIRMWARE_IMAGE_AUTHENTICATION.AuthInfo.Hdr.dwLength         = {dwLength:08X}'
```
Comment (spbrogan) [inline edk2toollib/uefi/wincert.py]: not right

Comment (corthon) [reply edk2toollib/uefi/wincert.py]: What are we talking about here?

Comment (corthon) [reply edk2toollib/uefi/wincert.py]: Ah... I see it.
Weird... wonder why that wasn't throwing exceptions all over the place.

Comment (corthon) [reply edk2toollib/uefi/wincert.py]: Wait... it's fine. The Encode was broken, though. What else are we talking about?

**Pull request merged**